[SaferCPP] Address issues for CSSPrimitiveValue

roberto-apple opened the pull request:
#### fe955121b3698d51e59c3e71ed0de1aec4c25eac
<pre>
[SaferCPP] Address issues for CSSPrimitiveValue
<a href="https://rdar.apple.com/149405072">rdar://149405072</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=291643">https://bugs.webkit.org/show_bug.cgi?id=291643</a>

Reviewed by Tim Nguyen.

Fix outstanding issues in CSSPrimitiveValue files.
Updated callers of CSSPrimitiveValue::equals() in CSSGridLineValue.cpp.

* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/css/CSSGridLineValue.cpp:
(WebCore::CSSGridLineValue::equals const):
* Source/WebCore/css/CSSGridLineValue.h:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::CSSPrimitiveValue::primitiveType const):
(WebCore::CSSPrimitiveValue::create):
(WebCore::CSSPrimitiveValue::resolveAsLengthDouble const):
(WebCore::CSSPrimitiveValue::doubleValue const):
(WebCore::CSSPrimitiveValue::doubleValueDeprecated const):
(WebCore::CSSPrimitiveValue::doubleValueDividingBy100IfPercentage const):
(WebCore::CSSPrimitiveValue::doubleValueDividingBy100IfPercentageDeprecated const):
(WebCore::CSSPrimitiveValue::stringValue const):
(WebCore::CSSPrimitiveValue::serializeInternal const):
(WebCore::CSSPrimitiveValue::equals const):
(WebCore::CSSPrimitiveValue::collectComputedStyleDependencies const):
(WebCore::CSSPrimitiveValue::customVisitChildren const):
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::TypeDeducingCSSValueMapper::operator unsigned short const):
(WebCore::TypeDeducingCSSValueMapper::operator int const):
(WebCore::TypeDeducingCSSValueMapper::operator unsigned const):
(WebCore::TypeDeducingCSSValueMapper::operator float const):
(WebCore::TypeDeducingCSSValueMapper::operator double const):
(WebCore::TypeDeducingCSSValueMapper::protectedNumericValue const):
(WebCore::CSSPrimitiveValue::convertToLength const):
(WebCore::TypeDeducingCSSValueMapper::numericValue const): Deleted.
* Source/WebCore/css/calc/CSSCalcValue.cpp:
(WebCore::CSSCalcValue::requiresConversionData const): Deleted.
* Source/WebCore/css/calc/CSSCalcValue.h:

Canonical link: <a href="https://commits.webkit.org/293925@main">https://commits.webkit.org/293925@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fa4614ff793b1333c37f309e5c00fd7381d89788

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100229 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19878 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10176 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105359 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50812 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102270 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20183 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28352 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76308 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33366 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103236 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15435 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90531 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56668 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15255 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8530 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50180 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/92887 "Build is in progress. Recent messages:") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85162 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8612 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107718 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/98836 "Build is in progress. Recent messages:") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27342 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20038 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85260 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27705 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86738 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84799 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21571 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29478 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7214 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/21213 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27279 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32512 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27090 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30406 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28649 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -1124,7 +1128,8 @@ String CSSPrimitiveValue::customCSSText(const CSS::SerializationContext& context
 bool CSSPrimitiveValue::equals(const CSSPrimitiveValue& other) const
 {
-    if (primitiveUnitType() != other.primitiveUnitType())
+    Ref otherRef = other;
```
Protection should happen at the callsite level, not in this method, the pattern of protecting in the method itself is not recommended.

roberto-apple replied:
To be clear, do you mean to update the signature to receive a Ref<const CSSPrimitive> parameter and/or find all the callers of CSSPrimitive::equals() and then create a smart object on the stack at the call site and then pass that in?

nt1m replied:
This: find all the callers of CSSPrimitive::equals() and then create a smart object on the stack at the call site and then pass that in

roberto-apple replied:
A follow-up question: If protection should happen for the callers of CSSPrimitiveValue::equals(), how would you solve this issue within the method itself (https://ews-build.s3-us-west-2.amazonaws.com/macOS-Safer-CPP-Checks-EWS/c446ef82-32329/scan-build-output/StaticAnalyzerRegressions/WTF/StaticAnalyzerReports/report-CSSPrimitiveValue.cpp-equals-86-9bfca0.html#EndPath) if you shouldn't protect as the code was in the original comment?

nt1m replied:
I think this is the fix: `return m_value.protectedCalc()->equals(*other.m_value.protectedCalc())`
or:

```
Ref calc = m_value.calc;
Ref otherCalc = *other.m_value.calc;
return calc->equals(otherCalc);
```

(In this case, the first option is preferred since the variables are just used once)

nt1m commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -257,7 +257,7 @@ CSSUnitType CSSPrimitiveValue::primitiveType() const
             return type;
 
-        return m_value.calc->primitiveType();
+        return protectedCssCalcValue()->primitiveType();
```
You could probably also apply the same pattern here of checking `if (Ref cssCalcValue = cssCalcValue())`

roberto-apple replied:
Ah that's right. Will update.

nt1m commented on Source/WebCore/css/CSSGridLineValue.cpp:
```diff
@@ -63,22 +63,22 @@ Ref<CSSGridLineValue> CSSGridLineValue::create(RefPtr<CSSPrimitiveValue>&& spanV
             return false;
-    } else if (other.m_spanValue)
+    } else if (otherSpanValue)
         return false;
```
I'd probably factor this repeated logic into a lambda:

```
auto equals = [](CSSPrimitiveValue* value, CSSPrimitiveValue* otherValue) {
    if ((!value && otherValue) || (value && !otherValue))
        return false;
    return (!value && !otherValue) || value->equals(otherValue);
};
```
then use it like this:
```
if (!equals(protectedSpanValue(), other.protectedSpanValue())
    return false;
if (!equals(protectedNumericValue(), other.protectedNumericValue())
    return false;
if (!equals(protectedGridLineName(), other.protectedGridLineName())
    return false;
```

cdumez commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -445,15 +445,15 @@ Ref<CSSPrimitiveValue> CSSPrimitiveValue::create(double value, CSSUnitType type)
     case CSSUnitType::CSS_NUMBER:
-        if (auto* result = valueFromPool(staticCSSValuePool->m_numberValues, value))
+        if (RefPtr result = valueFromPool(staticCSSValuePool->m_numberValues, value))
             return *result;
```
`return result.releaseNonNull();` would avoid ref-counting churn.

cdumez commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -445,15 +445,15 @@ Ref<CSSPrimitiveValue> CSSPrimitiveValue::create(double value, CSSUnitType type)
     case CSSUnitType::CSS_PERCENTAGE:
-        if (auto* result = valueFromPool(staticCSSValuePool->m_percentageValues, value))
+        if (RefPtr result = valueFromPool(staticCSSValuePool->m_percentageValues, value))
             return *result;
```
`return result.releaseNonNull();` would avoid ref-counting churn.

cdumez commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -445,15 +445,15 @@ Ref<CSSPrimitiveValue> CSSPrimitiveValue::create(double value, CSSUnitType type)
     case CSSUnitType::CSS_PX:
-        if (auto* result = valueFromPool(staticCSSValuePool->m_pixelValues, value))
+        if (RefPtr result = valueFromPool(staticCSSValuePool->m_pixelValues, value))
             return *result;
```
`return result.releaseNonNull();` would avoid ref-counting churn.

cdumez commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -553,8 +553,8 @@ Ref<CSSPrimitiveValue> CSSPrimitiveValue::createInteger(double value)
-    if (isCalculated())
-        return m_value.calc->requiresConversionData();
+    if (RefPtr calcValue = cssCalcValue())
+        return calcValue->requiresConversionData();
```
I think we can inline `requiresConversionData()` to avoid requiring a RefPtr here.

cdumez commented on Source/WebCore/css/CSSPrimitiveValueMappings.h:
```diff
@@ -97,39 +97,39 @@ class TypeDeducingCSSValueMapper {
 
     operator unsigned short() const
```
Is the `TypeDeducingCSSValueMapper` class dead code? I could not find any use of it. If it is truly unused, then I think we should remove it instead of updating it.

cdumez replied:
Note that the reason I wanted to check the usage of this class is because you made the data member a Ref:
```
Ref<const CSSValue> m_value;
```

This can introduce reference cycles and therefore leaks, depending on how this class is used. Without seeing the usage (since I couldn't find any), I'm unable to say if it's safe or not.

roberto-apple replied:
From what I have seen the sole place where it's used is here (https://github.com/WebKit/WebKit/blob/main/Source/WebCore/style/StyleBuilderConverter.h#L1383) where it is created as a temporary object returned by fromCSSValueDeducingType()

cdumez replied:
Oh that's right! it is used. It looks good then. Using a Ref for the data member should be fine in this case.

darinadler replied:
The `fromCSSValueDeducingType` function is also used in `BuilderConverter::convertScrollbarWidth`, but it shouldn’t be. The code there should instead just be `auto scrollbarWidth = fromCSSValue<ScrollbarWidth>(...)`.

roberto-apple replied:
Will update to use the `auto scrollbarWidth = fromCSSValue<ScrollbarWidth>(...)` suggestion.

Upon further investigation it looks like `fromCSSValueDeducingType` is used heavily in **DerivedSources/WebCore/StyleBuilderGenerated.cpp** as a result of being used in **process-css-properties.py**(https://github.com/WebKit/WebKit/blob/main/Source/WebCore/css/scripts/process-css-properties.py#L4096) (and numerous references here https://github.com/WebKit/WebKit/blob/main/Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp)

darinadler replied:
Some day, as a cleanup and to slightly speed up build time, we can move `fromCSSValueDeducingType` and `TypeDeducingCSSValueMapper` out of a more widely used header and into a header that is only included by `StyleBuilderGenerated.cpp` since that is the only file they should be used in.

roberto-apple replied:
Created https://bugs.webkit.org/show_bug.cgi?id=291812 as a follow-up task to track this recommendation.

cdumez commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -436,7 +436,7 @@ static CSSPrimitiveValue* valueFromPool(std::span<LazyNeverDestroyed<CSSPrimitiv
 {
-    if (auto* result = valueFromPool(staticCSSValuePool->m_numberValues, value))
+    if (RefPtr result = valueFromPool(staticCSSValuePool->m_numberValues, value))
         return *result;
```
`return result.releaseNonNull();` would avoid ref-counting churn.

nt1m commented on Source/WebCore/css/CSSGridLineValue.cpp:
```diff
@@ -63,22 +63,17 @@ Ref<CSSGridLineValue> CSSGridLineValue::create(RefPtr<CSSPrimitiveValue>&& spanV
-        if (!other.m_numericValue || !m_numericValue->equals(*other.m_numericValue))
-            return false;
-    } else if (other.m_numericValue)
+    if (!equals(protectedSpanValue().get(), other.protectedSpanValue().get()))
```
Does it compile without `.get()` ? I would drop if them if so

cdumez replied:
It wouldn't. a `Ref<T>` can implicitly get converted to a `T&` but a RefPtr<T>` cannot be implicitly converted to a `T*` with our current API.

roberto-apple committed:
[SaferCPP] Address issues for CSSPrimitiveValue
rdar://149405072
https://bugs.webkit.org/show_bug.cgi?id=291643

Reviewed by Tim Nguyen.

Fix outstanding issues in CSSPrimitiveValue files.
Updated callers of CSSPrimitiveValue::equals() in CSSGridLineValue.cpp.

* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/css/CSSGridLineValue.cpp:
(WebCore::CSSGridLineValue::equals const):
* Source/WebCore/css/CSSGridLineValue.h:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::CSSPrimitiveValue::primitiveType const):
(WebCore::CSSPrimitiveValue::create):
(WebCore::CSSPrimitiveValue::resolveAsLengthDouble const):
(WebCore::CSSPrimitiveValue::doubleValue const):
(WebCore::CSSPrimitiveValue::doubleValueDeprecated const):
(WebCore::CSSPrimitiveValue::doubleValueDividingBy100IfPercentage const):
(WebCore::CSSPrimitiveValue::doubleValueDividingBy100IfPercentageDeprecated const):
(WebCore::CSSPrimitiveValue::stringValue const):
(WebCore::CSSPrimitiveValue::serializeInternal const):
(WebCore::CSSPrimitiveValue::equals const):
(WebCore::CSSPrimitiveValue::collectComputedStyleDependencies const):
(WebCore::CSSPrimitiveValue::customVisitChildren const):
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::TypeDeducingCSSValueMapper::operator unsigned short const):
(WebCore::TypeDeducingCSSValueMapper::operator int const):
(WebCore::TypeDeducingCSSValueMapper::operator unsigned const):
(WebCore::TypeDeducingCSSValueMapper::operator float const):
(WebCore::TypeDeducingCSSValueMapper::operator double const):
(WebCore::TypeDeducingCSSValueMapper::protectedNumericValue const):
(WebCore::CSSPrimitiveValue::convertToLength const):
(WebCore::TypeDeducingCSSValueMapper::numericValue const): Deleted.
* Source/WebCore/css/calc/CSSCalcValue.cpp:
(WebCore::CSSCalcValue::requiresConversionData const): Deleted.
* Source/WebCore/css/calc/CSSCalcValue.h:

Canonical link: https://commits.webkit.org/293925@main
